### ISSUE TITLE
Not sent image file from 360Dialog WhatsApp

### ISSUE THREAD

**Issue by sultanlive** [chatwoot/chatwoot#3632]
**Not sent image file from 360Dialog WhatsApp**

**Comment by sojan-official** [comment]
@sultanlive  could you add more details. Did you mean that an image attachment was not received in chatwoot or vice versa?

**Comment by sultanlive** [comment]
@sojan-official I submitted an image from chatwoot, it didn't arrive to the client.
But the images come from clients

**Comment by sultanlive** [comment]
Have a news?

**Comment by sojan-official** [comment]
@sultanlive  can you add the following information? 

- are you using self-hosted installation / chatwoot cloud?
- if self-hosted, could you specify the chatwoot version 
- could you specify the image formate/size of the image 
- is the issue persistent for all uploads or its happening only at times.

**Comment by sultanlive** [comment]
I did not find in the source codes, the method using "https://waba.360dialog.io/v1/media"

It is mandatory to submit an image. Read more "https://docs.360dialog.com/whatsapp-api/whatsapp-api/media"

**Comment by sojan-official** [comment]
The API also supports sending a url where the attachment is hosted instead of using a media url

ref: https://github.com/chatwoot/chatwoot/blob/7b2ff2f112936c0905f4ef3ef608e1cae790ac38/app/models/channel/whatsapp.rb#L85

Could you confirm whether you are having the issue?

**Comment by sultanlive** [comment]
Yes, really did not notice this code. But I checked what Chatwoot sends to the file url.

The following request is sent:

curl --location --request POST 'https://waba.360dialog.io/v1/messages' \
--header 'D360-API-KEY:PgK7PJGcxfzUu6SeJICv7d7jAK' \
--header 'Content-Type: application/json' \
--data-raw '{
     "recipient_type": "individual",
     "to": "77476010836",
     "type": "image",
     image: {
        
     }
}'

the link to the file is a redirect, how can I fix this?
if you make a request directly to the file, then everything is OK

**Comment by sultanlive** [comment]
How to off redirects to file? 
Leave a link to a direct file

**Comment by sultanlive** [comment]
@sojan-official Hi, how disable redirects for attachment files?

**Comment by sultanlive** [comment]
@sojan-official  I sent from chatwoot cloud, file Also, the file is not stored directly, but by redirect.

https://app.chatwoot.com/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOVVnRHc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--33bf1bc286a14ea07d2c8a9d4ecc26f73cd8cb7f/Снимок%20экрана%202022-02-10%20в%2017.44.27.png

**Comment by sojan-official** [comment]
@sultanlive could you confirm whether the attachments for 360dialog isn't working for you even on chatwoot cloud?

**Comment by loama** [comment]
For me it isn't working on chatwoot self hosted but it is working in chatwoot cloud, using twilio.

Although I can receive media from whatsapp to chatwoot self hosted. And both send and receive from chatwoot website chat widget.

So the only issue seems to be on the post call made to twilio whatsapp for media.
Not sure of how to further debug that

**Comment by sultanlive** [comment]
For the test, I compared the responses from my server and from the Cloud.

![image](https://user-images.githubusercontent.com/47426878/159255526-faf8cfbb-5c20-4de3-b368-a06f7aaef7af.png)
Response differ
How to configure for nginx to work correctly?

**Comment by sultanlive** [comment]
My nginx config

**Comment by sultanlive** [comment]
![image](https://user-images.githubusercontent.com/47426878/159260804-1feb2b1c-b0a8-4504-9c8e-18bc853a47fc.png)

**Comment by loama** [comment]
@sojan-official I am using GCP, the problem seems to be the redirection.

When I test doing a direct HTTP request to Twilio I get this message

```javascript
{
    "code": 21620,
    "message": "Media urls: [my_url]//rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBWVE9IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--5ddeac0d527d25aad942d2d1860e1ea8943979a3/download.png are invalid. Please use only valid http and https urls",
    "more_info": "https://www.twilio.com/docs/errors/21620",
    "status": 400
}
```

(redacted url for security)

That happens with the heroku redirect links that chatwoot generates and send, but not with the gcp direct links.
To me, it seems like twilio doesn't like to follow redirections.

**Comment by sojan-official** [comment]
@loama  Thanks for debugging. We will look into this.

**Comment by sultanlive** [comment]
Tech support response 360 dialog
```
Hello.

I am not talking about the response. I am talking about a webhook call:

I have tested and this is what I got back:

{noformat}{
    "statuses": [
        {
            "errors": [
                {
                    "code": 1014,
                    "title": "Request for url http://flowsellbucket.hb.bizmrg.com/wdolkim7yn82hn3idjimo6z8d8xo?response-content-disposition=inline%3B filename%3D%22Screenshot from 2022-03-20 00-52-38.png%22%3B filename%2A%3DUTF-8%27%27Screenshot%2520from%25202022-03-20%252000-52-38.png&response-content-type=image%2Fpng&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=bt1VuP5pwTQpCnbVkkfskE%2F20220328%2Fhb%2Fs3%2Faws4_request&X-Amz-Date=20220328T170715Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&X-Amz-Signature=d12ace879230c48ed57aff875d3d89e968cc4e3be7510b0436fa8d97d959b7fe failed with error: 302 (Found)"
                }
            ],
            "id": "gBEGVRFRmZURAgkR8F9_p-MZsEo",
            "recipient_id": "551151999511",
            "status": "failed",
            "timestamp": "1648487235",
            "type": "message"
        }
    ],
    "incoming_messages": []
}{noformat}

I encourage you to try to upload the media using the {{/media}} endpoint and send media using the media ID you get back: [https://docs.360dialog.com/whatsapp-api/whatsapp-api/media|https://docs.360dialog.com/whatsapp-api/whatsapp-api/media|smart-link] 

Let me know if you can get it working or if you still require assistance.
```
